src/librustc_trans/trans/mir/did.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+//! Code for translating references to other items (DefIds).
+
+use rustc::front::map;
+use rustc::middle::ty::{self, Ty, TypeVariants};
+use rustc::middle::subst::Substs;
+use rustc::middle::const_eval;
+use rustc::middle::def_id::DefId;
+use trans::common::Block;
+use trans::base;
+use trans::expr;
+use trans::monomorphize;
+use trans::inline;
+
+use super::MirContext;
+use super::operand::{OperandRef, OperandValue};
+
+impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+    /// Translate reference to item.
+    pub fn trans_item_ref(&mut self,
+                          bcx: Block<'bcx, 'tcx>,
+                          ty: Ty<'tcx>,
+                          substs: &'tcx Substs<'tcx>,
+                          did: DefId)
+                          -> OperandRef<'tcx> {
+        match ty.sty {
```

Comment:
> Also, I haven't gone digging into trans, but is this code here a kind if port of code that exists elsewhere?

Nope. The `trans_fn_ref` below is a simplified “port”, but I haven’t found anything in old trans which would do matching like this (AFAICT it gathers such information from various other sources, esp. HIR).

> Perhaps we should just be including more information in the Literal MIR node? Or a Def (as I think you suggested on IRC)?

This approach makes it pretty easy to match whether enum/struct references should be translated as functions or not. I’m not sure other methods would allow checking for this as easily.

---

I’m willing to explore for better methods to match correct path, but won’t be able to dedicate much time for that in coming weeks.